File: Step 04 Binary Search/Step 4.3 Find Answers by BS in Search Space/Find Nth root of M/Find Nth root of M.cpp
class Solution{
	public:
	long long power(int base, int exp) {
        long long result = 1;
        for (int i = 0; i < exp; i++) {
            result *= base;
            if (result > INT_MAX) {  // Early termination if result exceeds integer bounds
                return INT_MAX;
            }
        }
        return result;
    }
    
    int NthRoot(int n, int m)
    {
        // Code here.
        if (m == 0 || m == 1) {
            return m;
        }
        
        int low = 1, high = m;
        while (low <= high) {
            int mid = low + (high - low) / 2;
            long long mid_pow = power(mid, n);
            
            if (mid_pow == m) {
                return mid;
            }
            else if (mid_pow < m) {
                low = mid + 1;
            }
            else {
                high = mid - 1;
            }
        }
        return -1;
	}  
};

File: Step 04 Binary Search/Step 4.3 Find Answers by BS in Search Space/Find Nth root of M/Find Nth root of M.java


//User function Template for Java

class Solution
{
    public int NthRoot(int n, int m)
    {
        // code here
        int low = 1;
        int high = m;
        while (low <= high) {
            int mid = low + (high - low) / 2;
            int poww = (int)Math.pow(mid, n);
            if (poww == m) return mid;
            else if (poww < m) low = mid + 1;
            else high = mid - 1;
        }
        return -1;
    }
}